File: detector/random_ferns/RandomFerns.py
import numpy as np
from numpy import random as rnd


class RandomFerns:
    def __init__(self, pool_size=5, fern_size=4, num_feats=5, fern_depth=3):
        self.__pool_size = pool_size
        self.__fern_size = fern_size
        self.__num_feats = num_feats
        self.__fern_depth = fern_depth

        self.__data = None

    @property
    def pool_size(self):
        return self.__pool_size

    @pool_size.setter
    def pool_size(self, size):
        self.__pool_size = size

    @property
    def fern_size(self):
        return self.__fern_size

    @fern_size.setter
    def fern_size(self, size):
        self.__fern_size = size

    @property
    def num_feats(self):
        return self.__num_feats

    @num_feats.setter
    def num_feats(self, num):
        self.__num_feats = num

    @property
    def fern_depth(self):
        return self.__fern_depth

    @fern_depth.setter
    def fern_depth(self, depth):
        self.__fern_depth = depth

    @property
    def data(self):
        return self.__data

    @data.setter
    def data(self, data):
        self.__data = data

    def compute(self, r, m, s, c):
        # r: num. ferns parameters (pool size)
        # m: num. binary features
        # s: fern size
        # c: num. image channels (e.g color channels
        self.pool_size = r
        self.num_feats = m
        self.fern_size = s
        self.fern_depth = c

        print('ps - {}, nf - {}, fs - {}, fd - {}'.format(self.pool_size, self.num_feats, self.fern_size,
                                                          self.fern_depth))

        temp = []

        for i in range(r):
            for j in range(m):
                ua = np.floor(s * rnd.random())
                va = np.floor(s * rnd.random())
                ca = np.floor(c * rnd.random())

                ub = np.floor(s * rnd.random())
                vb = np.floor(s * rnd.random())
                cb = np.floor(c * rnd.random())

                row = [ua, va, ca, ub, vb, cb]
                temp.append(row)

        self.data = np.vstack(temp)
        self.data = self.data.reshape(r, m, 6)
        return self.data

    def print(self):
        print('\n*********************************************************')
        print('* Random Fern Parameters : ')
        print('* num. ferns parameters: {}'.format(self.pool_size))
        print('* num. binary features: {}'.format(self.num_feats))
        print('* fern size: {} x {}'.format(self.fern_size, self.fern_size))
        print('* fern depth: {}'.format(self.fern_depth))
        print('*********************************************************\n')

#
# if __name__ == '__main__':
#     rf = RandomFerns()
#     comp = rf.compute(10, 10, 10, 3)
#     print(comp)
#     rf.print()

File: src/randomFerns/RandomFerns.cpp
#include "RandomFerns.h"

RandomFerns::RandomFerns() {
    // initialize default parameters
    this->poolSize = 5;  // num. ferns parameters (R)
    this->fernSize = 4;  // fern size (S)
    this->numFeats = 5;  // num. binary features per fern (M)
    this->fernDepth = 3;  // fern depth (num. image channels (C))
}

RandomFerns::RandomFerns(int poolSize, int numFeats, int fernSize, int fernDepth) {
    this->poolSize = poolSize;
    this->numFeats = numFeats;
    this->fernSize = fernSize;
    this->fernDepth = fernDepth;
}

RandomFerns::~RandomFerns() = default;

// get num. ferns parameters (R)
int RandomFerns::get_pool_size(){
    return this->poolSize;
}

// get num. binary features (M)
int RandomFerns::get_num_features(){
    return this->numFeats;
}

// get fern size (S)
int RandomFerns::get_fern_size(){
    return this->fernSize;
}

// get fern depth (image channels (C))
int RandomFerns::get_fern_depth(){
    return this->fernDepth;
}

// get ferns parameters data
cv::Mat RandomFerns::get_data(){
    return this->data;
}

py::array_t<unsigned char> RandomFerns::get_data_np() {
    py::array_t<unsigned char > result = py::array_t<unsigned char>({ this->data.rows, this->data.cols, 3}, this->data.data);
    return result;
}

// compute random ferns
void RandomFerns::compute(int R, int M, int S, int C){
    // inputs:
    // R: num. ferns parameters (pool size)
    // M: num. binary features
    // S: fern size
    // C: num. image channels (e.g color channels)
    // variables
    int ua,va,ca,ub,vb,cb;

    // set random ferns values
    this->poolSize = R;  // num. ferns parameters
    this->numFeats = M;  // num. binary features
    this->fernSize = S;  // spatial fern size
    this->fernDepth = C;  // fern depth (num. image channels)

    // allocate data
    this->data = cv::Mat(R, M, CV_8UC(6));

    // pointer to ferns data
    auto *dataPtr = (unsigned char*)(data.data);

    // random ferns parameters
    for (int r = 0; r<R; r++ ){
        for (int m = 0; m<M; m++){
            // feature a coordinates
            ua = floor(S*((double)rand()/RAND_MAX));
            va = floor(S*((double)rand()/RAND_MAX));
            ca = floor(C*((double)rand()/RAND_MAX));

            // feature b coordinates
            ub = floor(S*((double)rand()/RAND_MAX));
            vb = floor(S*((double)rand()/RAND_MAX));
            cb = floor(C*((double)rand()/RAND_MAX));

            // save
            *(dataPtr + r*M*6 + m*6 + 0) = ua;
            *(dataPtr + r*M*6 + m*6 + 1) = va;
            *(dataPtr + r*M*6 + m*6 + 2) = ca;
            *(dataPtr + r*M*6 + m*6 + 3) = ub;
            *(dataPtr + r*M*6 + m*6 + 4) = vb;
            *(dataPtr + r*M*6 + m*6 + 5) = cb;
        }
    }
}

void RandomFerns::print() {
    std::cout << "\n*********************************************************" << std::endl;
    std::cout << "* Random Fern Parameters : " << std::endl;
    std::cout << "* num. ferns parameters -> " << this->poolSize << std::endl;
    std::cout << "* num. binary features -> " << this->numFeats << std::endl;
    std::cout << "* fern size ->  " << this->fernSize << "x" << this->fernSize << std::endl;
    std::cout << "* fern depth ->  " << this->fernDepth << std::endl;
    std::cout << "*********************************************************\n" << std::endl;
}
